### DISCUSSION TITLE
Craft 4 / rewrite

### DISCUSSION THREAD

**Discussion by ostark** [ostark/upper#65] [Ideas]
### Restart

I noticed people still rely on this plugin. TBH, I didn't pay much attention to it within the previous months or even years. 
It seems to be stable. 

However, I am fairly unhappy with the codebase. This is one of the first Craft plugins [I started working on 5 years ago](https://github.com/ostark/upper/commit/2b942a976680efa3cd4de56f24730fbbd3f64bac) - today it feels like a mess 🤛🏼. 

Ping: 
@markhuot @tomdavies @larsboldt @blasvicco @bartdigitalpulse @Hyra @L3P3 @taylordaughtry @devinellis @toddpadwick @timkelty 
(your input is highly appreciated)

### Things I consider to add/change

* Unit tests
* Strict Types 
* Config structure 

### Features I consider to add

* Multi-layer caching, including local
* Better multi-site support https://github.com/ostark/upper/issues/47 
* CloudFront Support https://github.com/ostark/upper/issues/8

**Comment by timkelty** [comment]
@ostark Is there anything you can think of (events, etc) that Craft could provide to make Upper's job easier?

**Reply by ostark** [reply]
maybe this? https://github.com/craftcms/cms/discussions/11171

**Reply by ostark** [reply]
Not sure if I want to use `Element::getCacheTags()` eventually.
If I would, I had to rename them to a shorter string to keep the headers lightweight.

**Comment by timkelty** [comment]
Just made this issue that would be nice to consider: https://github.com/ostark/upper/issues/66

**Comment by markhuot** [comment]
A few quick thoughts,

1. yes, I still use (and love) this plugin. I like it because it's so light weight and does one thing and does it well. Your first bullet, about "including local" caching worries me because I think it deviates from what makes this plugin so good.
2. I've forked this plugin for local use because I wanted more control over the tags that get printed as well as the tags that get purged. It would be nice if you could have drivers on both sides. Right now you have drivers for clearing caches but no drivers for determining tags. E.g., the way I'm using globals I don't really want them tagged on every page. Changing a global is a big deal on a few of my sites and I don't really want to blow every page on the site every time it changes. That's a more considered thing where I'll make the change in Craft and blow the pages via the CDN UI over night/off peak.
3. Happy to provide whatever help/assistance you need moving forward and answer any more questions you may have.

**Reply by timkelty** [reply]
> Your first bullet, about "including local" caching worries me because I think it deviates from what makes this plugin so good.

Agreed! IMO headers are the only way to go, and think it might me worth considering limiting scope to that.

**Reply by markhuot** [reply]
Especially with `stale-while-revalidate` on Cloudflare you can simulate two layers with a single service. Origin requests get queued up at the edge (or served stale content) while only a single origin request is made.

**Reply by ostark** [reply]
Thanks for your feedback so far!

* @markhuot True, the "local caching" might add some complexity. >> It's not set in stone.
* @timkelty IMO headers are the only way to go >> I don't disagree here
* @markhuot ... no drivers for determining tags >> I'll keep this in mind

**Comment by ostark** [comment]
Note to myself:
https://github.com/craftcms/cms/discussions/11171#discussioncomment-3041733

**Comment by khalwat** [comment]
This PR may be helpful: https://github.com/craftcms/cms/pull/11617
